python/lsst/daf/butler/core/timespan.py

```diff
@@ -35,31 +35,116 @@
 
 
 class Timespan(Generic[T], tuple):
-
-    def __new__(cls, begin: T, end: T):
+    """A generic 2-element named tuple for time intervals.
+
+    Parameters
+    ----------
+    begin : ``T``, optional
+        Minimum timestamp in the interval (inclusive).  `None` is interpreted
+        as -infinity.
+    end : ``T``, optional
+        Maximum timestamp in the interval (inclusive).  `None` is interpreted
+        as +infinity.
+
+    Notes
+    -----
+    This class is generic because it is used for both Python literals (with
+    ``T == astropy.time.Time``) and timestamps in SQLAlchemy expressions
+    (with ``T == sqlalchemy.sql.ColumnElement``), including operations between
+    those.
+
+    Python's built-in `collections.namedtuple` is not actually a type (just
+    a factory for types), and `typing.NamedTuple` doesn't support generics,
+    so neither can be used here (but also wouldn't add much even if they
+    could).
+    """
+    def __new__(cls, begin: Optional[T], end: Optional[T]):
         return tuple.__new__(cls, (begin, end))
 
-    def overlaps(self, other, ops=operator):
-        return ops.not_(ops.or_(self.end < other.begin, self.begin > other.end))
+    def overlaps(self, other, ops=operator) -> Any:
+        """Test whether this timespan overlaps another.
+
+        Parameters
+        ----------
+        other : `Timespan`
+            Another timespan whose begin and end values can be compared with
+            those of ``self`` with the ``>=`` operator, yielding values
+            that can be passed to ``ops.or_`` and/or ``ops.and_``.
+        ops : `Any`, optional
+            Any object with ``and_`` and ``or_`` boolean operators.  Defaults
+            to the Python built-in `operator` module, which is appropriate when
+            ``T`` is a Python literal like `astropy.time.Time`.  When either
+            operand contains SQLAlchemy column expressions, the
+            `sqlalchemy.sql` module should be used instead.
+
+        Returns
+        -------
+        overlaps : `Any`
+            The result of the overlap.  When ``ops`` is `operator`, this will
+            be a `bool`.  If ``ops`` is `sqlachemy.sql`, it will be a boolean
+            column expression.
+        """
+        return ops.and_(
+            ops.or_(self.end == None, other.begin == None, self.end >= other.begin), # noqa E711
```

Comment:
Thanks; I always forget which to use.  But while fixing this I discovered a handy trick to remind me in the future: if you get it right, flake8 will _also_ tell you if you only put one space before the comment instead of two, and if you get it wrong, it won't.